websocket/connection/upgrader.go

```diff
@@ -0,0 +1,126 @@
+package connection
+
+import (
+	"fmt"
+	"net/http"
+	"raccoon/logger"
+	"raccoon/metrics"
+	pb "raccoon/websocket/proto"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/websocket"
+)
+
+type Upgrader struct {
+	gorillaUg         websocket.Upgrader
+	Table             *Table
+	pongWaitInterval  time.Duration
+	writeWaitInterval time.Duration
+	connIDHeader      string
+	connGroupHeader   string
+}
+
+type UpgraderConfig struct {
+	ReadBufferSize    int
+	WriteBufferSize   int
+	CheckOrigin       bool
+	MaxUser           int
+	PongWaitInterval  time.Duration
+	WriteWaitInterval time.Duration
+	ConnIDHeader      string
+	ConnGroupHeader   string
+}
+
+func NewUpgrader(conf UpgraderConfig) *Upgrader {
+	var checkOriginFunc func(r *http.Request) bool
+	if conf.CheckOrigin == false {
+		checkOriginFunc = func(r *http.Request) bool {
+			return true
+		}
+	}
+	return &Upgrader{
+		gorillaUg: websocket.Upgrader{
+			ReadBufferSize:  conf.ReadBufferSize,
+			WriteBufferSize: conf.WriteBufferSize,
+			CheckOrigin:     checkOriginFunc,
+		},
+		Table:             NewTable(conf.MaxUser),
+		pongWaitInterval:  conf.PongWaitInterval,
+		writeWaitInterval: conf.WriteWaitInterval,
+		connIDHeader:      conf.ConnIDHeader,
+		connGroupHeader:   conf.ConnGroupHeader,
+	}
+}
+
+func (u *Upgrader) Upgrade(w http.ResponseWriter, r *http.Request) (Conn, error) {
+	identifier := NewConnIdentifier(r.Header, u.connIDHeader, u.connGroupHeader)
+	logger.Debug(fmt.Sprintf("%s connected at %v", identifier, time.Now()))
+
+	conn, err := u.gorillaUg.Upgrade(w, r, nil)
+	if err != nil {
+		metrics.Increment("user_connection_failure_total", fmt.Sprintf("reason=ugfailure,conn_group=%s", identifier.Group))
+		return Conn{}, fmt.Errorf("failed to upgrade %s: %v", identifier, err)
+	}
+	if u.Table.Exists(identifier) {
+		duplicateConnResp := createEmptyErrorResponse(pb.Code_MAX_USER_LIMIT_REACHED)
+
+		conn.WriteMessage(websocket.BinaryMessage, duplicateConnResp)
+		conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(1008, "Duplicate connection"))
+		metrics.Increment("user_connection_failure_total", fmt.Sprintf("reason=exists,conn_group=%s", identifier.Group))
+		conn.Close()
+		return Conn{}, fmt.Errorf("disconnecting %s: already connected", identifier)
+	}
+	if u.Table.HasReachedLimit() {
```

Comment:
Maybe this check can be a part of the Store() method itself. Store can return an error if the limit has reached. This will reduce an extra function call here.